### ISSUE TITLE
need for aliases if all component names are unique?

### ISSUE THREAD

**Issue by nickjackson** [componentjs/builder.js#117]
Is there a possibility of removing aliases from the build if all of the components being used have names that are unique?

Also, I am in a position where removing the github namespace wouldn't cause any collisions either, so removing the username portion from any `require.register` calls would be helpful too.

Is there an option we can add that would achieve something like this?

**Comment by tj** [comment]
technically we could get rid of all of them and just compile a `modules` object that holds them all, and replace the `require()` calls with `modules['component/emitter']` etc, just haven't gone down that road yet but as far as the aliases taking up space etc that's all an implementation detail

**Comment by pgherveou** [comment]
+1
that would be a great enhancement
at the moment, the aliases takes more than half of the size of our 2MB build
Are there any tips we can use at the moment to reduce the fat?

**Comment by nickjackson** [comment]
Its a little lame but I am overriding the builder prototypes `.buildAlias()` and `.buildScripts()` to do my own thing.

**Comment by tj** [comment]
more than half the size?? that doesn't sound right at all, especially when gzipped, gzip should greatly reduce stuff like that since it's so redundant, definitely something we can improve though

**Comment by pgherveou** [comment]
gzipping works great, I was just checking size on disk. I am also experimenting storing and loading js code from localstorage for offline access and all this extra strings break maximum quotas sometimes. @nickjackson do you have a fork I could check for a tmp solution?

**Comment by tj** [comment]
yeah the cache aspect is definitely valid, I haven't seen that level of bloat though

**Comment by ronkorving** [comment]
Duplicate of #55 ?

**Comment by micky2be** [comment]
It is true that the `require.alias` are a bit crazy in builds.
In the mobile app I'm working on the size of the build is 3MB (non-minified), and 30% of it comes from all those `require.alias`.
Also using localStorage for caching but in any cases if we could have 1MB less to load that would more than awesome.

**Comment by tj** [comment]
30%? yikes, that definitely sounds not good, simplest thing to do for now would be to remove the alias mechanism and replace the `require()s` with a ref to the module like `modules['component-tip/index.js']` like I mentioned above, I can try and get on this soon.

**Comment by micky2be** [comment]
Would be nice yes 
Thank you

**Comment by jonathanong** [comment]
i'm at 40-50% aliases minified and pre-gzip :( got worse after i spent a day `boot`ing my app. but there's less than a 5% different in the final gripped size.

**Comment by aroc** [comment]
@visionmedia Is there anything else I need to do to replace the alias mechanism other than remove `require()` calls and replace them with `module[]` calls?

**Comment by tj** [comment]
@aroc nope that should be it, if I have a minute today I'll see if I can hack it in

**Comment by aroc** [comment]
@visionmedia Awesome! Thanks for the speedy response. You're the best maintainer on Github!

**Comment by tj** [comment]
I guess the one downside of doing this is we'd probably need two passes, otherwise the "lazy" nature of require() wouldn't work to well. For example if you have ./foo/index.json instead of ./foo/index.js we can't statically resolve that until all of "foo" is resolved which may or may not be before whatever is requiring it

**Comment by tbrebant** [comment]
@visionmedia I am looking forward this change also: in the project I'm currently working on, the size ratio between the require.alias(es) and the application code itself turned to something incredible. I hope you will find some time to work on this soon ;)

**Comment by ronkorving** [comment]
On my end, I've now "solved" the duplication problem without touching component-builder. Just to inspire some people who have a controlled build process like I do, here's the simple solution I apply. I don't use `builder.build()` anymore. I call builder's type-specific build methods and chain the results together just like `build` does. In my case that means (simplified for clarity):

``` js
builder.buildScripts(function (error, scripts) { .. });
builder.buildAliases(function (error, aliases) { .. });

var result = [];
result.push(require('component-require'));
result.push(scripts);
result.push(compressAliases(aliases));
var out = result.join('\n');

// this function removes all duplicate aliases and returns the remaining ones sorted,
// so that we guarantee the output to be predictable and friendly to readers.
function compressAliases(aliases) {
    if (typeof aliases !== 'string') {
        return '';
    }

    aliases = aliases.split('\n');

    var oal = {};

    for (var i = 0, len = aliases.length; i < len; i++) {
        oal[aliases[i]] = true;
    }

    return Object.keys(oal).sort().join('\n');
}
```

**Comment by micky2be** [comment]
:+1:

**Comment by bmcmahen** [comment]
+1 I'm running into this as well with my latest app that makes heavy use of `local` components and a main `boot` for loading. Kinda scary how big it's getting, although it's definitely alleviated by gzip.

**Comment by tj** [comment]
ran ^ on two of my larger projects and builds are fine, if some of you guys could confirm using component HEAD that would be great!

**Comment by pgherveou** [comment]
My builds are broken because it does not seem to support to main anymore
but it looks like you have already spotted that

**Comment by tj** [comment]
i forgot to uncomment those tests haha, sorry, will fix that today

**Comment by ronkorving** [comment]
For those who still depend on component-builder with aliases, I have augmented the workaround I pasted above with one more step, which introduces a handled-aliases list and skips any processing of aliases already handled (which means less code running and less file i/o). On one of the projects I'm involved in, build times got roughly 20x faster.

``` js
var orgBuildAliases = Builder.prototype.buildAliases;
var handledAliases = {};

Builder.prototype.buildAliases = function (callback) {
    if (handledAliases[this.dir]) {
        return callback(null, '');
    }

    handledAliases[this.dir] = true;

    orgBuildAliases.call(this, callback);
};
```

I still post-process with the code snippet from my earlier comment in this thread. The reason for that is drop whitespace and the double entries the "main" field of component.json (since that file will also occur in "scripts").
